### DISCUSSION TITLE
Allow lamps to be daisy chained by changing their power pins to be 2 wire junction box power pins

### DISCUSSION THREAD

**Discussion by kwogger** [FakeFishGames/Barotrauma#7839] [Ideas and suggestions]
A common issue for sub building has the requirement for large amounts of junction boxes due to lamps. This leads to the following suboptimal solutions:
1. Use of multiple relays and/or chained relays for more power connections for lamps. This however removes undervolting light flickering effects and may also lead to power grid instability when the relays are chained.
2. Use of hidden invulnerable junction boxes. Conceptually, it's just turning a junction box into a relay in 1. This end up hiding some of the wiring such that players cannot modify things they otherwise could have (e.g. players cannot disconnect wires as the junction boxes are hidden).
3. Just add a lot more junction boxes. This leads to issues where there's too many for players to repair. Making some of them invulnerable/more resilient isn't a great solution as players could exploit this to rewire it so it other machines using the more robust junction boxes instead of just lamps.
4. Wire a power detection circuit into the grid (usually a hidden battery with a synthetic load and some logic), set the lights to use no power, and control the lights on/off from that with locked wiring. Similar issues to 1 and 2, except can be used for replicating lighting effects.

These issues can be remedied by reducing the number of connections required for lamps by changing their power pin behaviour to allow for chaining (i.e. like Christmas lights). This can be accomplished by having their power pin setup like a junction box, but limited to just 2 pins.

Further gameplay enhancements are also added from this for traitor gameplay, as this will allow clever strategic unwiring to disable strings of lights.

**Comment by ignomen** [comment]
I like this but I'm leery of players using lamps as go-betweens for non-lamps in the submarine. Perhaps it would be better to have them only output power/be connectable to other lamps?

Additionally, if lamps could be subject to the overvoltage mechanic just as junction boxes are, this could be made really interesting. At an overvoltage of 1.1 for example, with a high default fire chance, you could have lamps exploding all across the submarine during a meltdown or some other scenario. Maybe keep lamps waterproof though.

**Reply by cshingle** [reply]
@ignomen @MaaiG Take a look at the comments on this PR. It is for a rework of the power system. These changes have been discussed.
https://github.com/Regalis11/Barotrauma/pull/8184

**Comment by MaaiG** [comment]
Yep. The problem is there and it’s not just with lamps.
Furthermore еhere are also problems with the relay.
Honestly, I don’t understand why these relays don’t work and the electricity doesn’t run past the first relay.
![Image 001_23-39](https://user-images.githubusercontent.com/93348499/146655017-c19cbb6e-b655-43e8-b1bf-adb9cd6b7537.png)

Sometimes the relays work as a capacitor. The first switching on is delayed. Then everything works fine.
Well, something’s not right here.

**Reply by gerakis** [reply]
I made a bug report about this involving relays and diving suit racks. Is it also happening with lights?

**Reply by gerakis** [reply]
I just tested it. It is happening with everything. Its a problem with daisy chaining relays together. Some bug introduced in the new patch.

**Reply by MaaiG** [reply]
> I made a bug report about this involving relays and diving suit racks. Is it also happening with lights?

[1-Test_Relay-Lamp.zip](https://github.com/Regalis11/Barotrauma/files/7745604/1-Test_Relay-Lamp.zip)

**Reply by cshingle** [reply]
@MaaiG correctly reported this as a bug. The screenshot of the issue was helpful in resolving it. It sounds like the fix will be in the upcoming version.
https://github.com/Regalis11/Barotrauma/issues/7864

**Comment by Tsunder** [comment]
I don't believe in 3. becaues it takes a LOT of lights to be noticably more than there are available engineers

even the largest ship, the berillia, is designated for 12+ players, can easily fit  2 - 4 electrians aboard to cover the ~20 JBs.

**Comment by r4d6** [comment]
Well, one junction box can connect to 4 relays, since it has 5 connection and it need one to get power from.
Each relay can power 5 lamps on their own.
So a single junction can power 20 lamps.

I honestly doubt any sub would have more than 3 junction boxes dedicated to lighting, not to mention that the sub builder could just wire the lamps with relays alongside the equipment that use less than 1 kW of power.

**Comment by gerakis** [comment]
Maybe just change how relays work so that relay connected lights can still flicker somehow.

**Comment by MaaiG** [comment]
> even the largest ship, the berillia, is designated for 12+ players

>So a single junction can power 20 lamps.

That’s what you think )) _(But what difference does it make? Relays still work illogically!)_

I never use vanilla submarines.
I make my own and they’re big. My current submarine is 73x26 meters.
I optimize logical systems (using fewer components if possible). I try to reduce the number of objects used twice if it is possible to use them only once, I’m minimizing the number of shadow lamps without detriment to design and etc. My docking ports are smarter than the vanilla sub with its bots.
I’m just getting started and I haven’t given "beauty" yet:
![Image 001_15-51](https://user-images.githubusercontent.com/93348499/146772984-30635b2e-3897-4b9a-a0b1-a16963cdb105.png)
**And that doesn’t include: The Shuttle, The Lifeboats, The drones, The lifting equipment, etc!**

I don’t have a problem with lighting, but I think there’s exist a problem with relay logic!

>Maybe just change how relays work so that relay connected lights can still flicker somehow.

Yep.

**Reply by Oukido** [reply]
Do you use those lights for actual illumination? You can reduce the need for that drastically by changing the ambiance lighting of the hull.
I use this often in conjuction with emergency lights which are basically always on (but do not really have to illuminate anything).
So real lights only have to add a little bit of light value ontop of that ambience light on spots that need to be well lit. If you use Red emergency lighting in the hull and emergency lights you might also want to adjust the color of the light to add up to a good color when it is toggled on

**Reply by MaaiG** [reply]
> Do you use those lights for actual illumination?

Yes, but some are used to indicate. If the light from the source does not spread outside the `hull` I turn off its shadows.
> You can reduce the need for that drastically by changing the ambiance lighting of the hull.

I already did that.
> I use this often in conjuction with emergency lights

Thank you. Usually I try to make the emergency lights are only turned on when the main light is switched off.

I gave this an example of a lot of lamps, not because it’s a big problem. I just answered the question, "Who might need so many lamps."
I thing the problem is in the "relay" component. And this is a topic that tries to partially solve the problem with lighting. But the main problem is the "relay" component IMHO.
